include/envoy/router/internal_redirect.h

```diff
@@ -0,0 +1,51 @@
+#pragma once
+
+#include "envoy/config/typed_config.h"
+#include "envoy/stream_info/filter_state.h"
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Router {
+
+/**
+ * Used to decide if an internal redirect is allowed to be followed based on the target route.
+ */
+class InternalRedirectPredicate {
+public:
+  virtual ~InternalRedirectPredicate() = default;
+
+  /**
+   * A FilterState is provided so that predicate implementation can use it to preserve state across
+   * internal redirects.
+   *
+   * @return whether the route specified by target_route_name is allowed to be followed. Any
+   *         predicate returning false will prevent the redirect from being followed, causing the
+   *         response to be proxied to the downstream.
```

Comment:
optional nit
to the downstream -> downstream

Reply:
done.